src/app/action-items/action-items.component.ts

```diff
@@ -1,38 +1,67 @@
-import { ActionTasksElement } from './../services/tasksService/tasks.service';
+import { AddItemComponent } from './../add-item/add-item.component';
 import { Component, OnInit, OnDestroy } from '@angular/core';
 import { map } from 'rxjs/operators';
-import { TasksService } from '../services/tasksService/tasks.service';
+import { TasksService, ActionItem } from '../services/tasksService/tasks.service';
 import { DaysLeftToDeadlineService } from '../services/daysLeftToDeadlineService/days-left-to-deadline.service';
+import { MatDialog, MatDialogConfig } from '@angular/material/dialog';
+import { Subscription } from 'rxjs';
 
-export interface ActionTasksElementMapped extends ActionTasksElement {
-  dueDay: number;
+export interface ActionItemMapped extends ActionItem {
+  dueDay?: number;
 }
 
 @Component({
   selector: 'app-action-items',
   templateUrl: './action-items.component.html',
   styleUrls: ['./action-items.component.css']
 })
-export class ActionItemsComponent implements OnInit {
-  dataSource: ActionTasksElementMapped[];
-  loading = true;
-  currentDate = new Date();
-  constructor(private tasksService: TasksService, private daysCountService: DaysLeftToDeadlineService) {}
+export class ActionItemsComponent implements OnInit, OnDestroy {
+  dataSource: ActionItemMapped[];
+  isloadingActionItems = false;
+  private currentDate: Date = new Date();
+  private subscription: Subscription;
+  constructor(
+    private tasksService: TasksService,
+    private daysCountService: DaysLeftToDeadlineService,
+    private matDialog: MatDialog
+  ) {}
   ngOnInit() {
-    this.tasksService
-      .getAllTasks()
+    this.retrieveActionItems();
+  }
+  retrieveActionItems() {
+    this.isloadingActionItems = true;
+    this.subscription = this.tasksService
+      .getAllItems()
       .pipe(
         map(items => {
-          const mappedActionItems = items.map(item => {
-            const dueDayCounted = this.daysCountService.daysLeftToDeadline(item.dueDate, this.currentDate);
+          const mappedActionItems: ActionItemMapped[] = items.map(item => {
+            const dueDayCounted = item.dueDate
+              ? this.daysCountService.daysLeftToDeadline(item.dueDate, this.currentDate)
+              : undefined;
             return { ...item, dueDay: dueDayCounted };
           });
           return mappedActionItems;
         })
       )
       .subscribe(tasks => {
         this.dataSource = tasks;
-        this.loading = false;
+        this.isloadingActionItems = false;
+      });
+  }
+  openDialog() {
+    const dialogConfig = new MatDialogConfig();
+    dialogConfig.data = { width: '470px', height: 'auto' };
+    this.matDialog
+      .open(AddItemComponent, dialogConfig.data)
+      .afterClosed()
+      .subscribe(item => {
+        if (item !== undefined) {
+          this.retrieveActionItems();
+        }
+        this.isloadingActionItems = false;
```

Comment:
There is no need to set this here, it is handled directly by `retrieveActionItems` function - in this implementation loading spinner for action items will be hidden even when we are still waiting to get action items from the service which is not correct behaviour.